### ISSUE TITLE
Poor performances when typing in rectangle selection

### ISSUE THREAD

**Issue by nippur72** [icsharpcode/AvalonEdit#58]
On a file with about 1000 lines of text and syntax highlighting turned on, making a rectangle selection with more ~15 lines makes editing inside the selection very slow, dropping to something like 1 character per second.

**Comment by zaksmolen** [comment]
I am still noticing this with current versions of AvalonEdit. I even turned off syntax highlighting and collapsing and even so, when I have a decent size block selection and try to type sometimes it takes a second per character.

Anyone have any ideas about fixes or optimizations? We unfortunately have a lot of people doing block selecting in our application...

**Comment by zaksmolen** [comment]
I have noticed that some of the things mentioned in #9 do help. Turning off space showing does make it a little faster, but we need spaces shown, and also it really seems to happen more with block select.

**Comment by goswinr** [comment]
@siegfriedpammer could this be fixed by using [`Document.BeginUpdate() `](http://avalonedit.net/documentation/html/d2f40c5c-7f5f-8a07-208e-702288f4e19f.htm) and `Document.EndUpdate()` 
Or are you doing this already?
Could you point me at the relevant code, so that I can maybe try a PR ...

**Comment by zaksmolen** [comment]
I would be happy to take a look at this as well if someone could point me to the relevant area in the code. We finally were able to update our VS version so I can actually give fixing this a try!
Our users are literally copy/pasting in/out of WordPad or NP++ in order to do block selection and typing because that's faster than trying to do it in the editor.

**Comment by zaksmolen** [comment]
@goswinr I'm looking at the code and they do seem to have that area of code wrapped with `using (textArea.Document.RunUpdate())` (so it does begin/end update), so that's not going to change anything.

I did some benchmark testing on my brand new and decently high-spec computer, timing this section of the code:
```
foreach (SelectionSegment lineSegment in this.Segments.Reverse()) {
	ReplaceSingleLineText(textArea, lineSegment, newText, out insertionLength);
	totalInsertionLength += insertionLength;
	firstInsertionLength = insertionLength;
}
```

~100 lines took consistently 1.5-4 seconds
~500 lines took consistently almost a full minute

I do see that there's a call to `RaiseAutomationEvent(AutomationEvents.TextPatternOnTextSelectionChanged);` for every line that's changed, but I'm looking into the code more to see if that's a possible area of the slowdown.

**Comment by zaksmolen** [comment]
Just to add more as I debug, 90+% of the time is in `TextArea.OnDocumentChanged()` and almost all of that is the `selection.UpdateOnDocumentChange(e)` line. So it's the `UpdateOnDocumentChange()` method of `RectangularSelection`.

I'm not sure why the selection needs to be updated after each line of a change, and not just at the end. I.e. it seems like this update should be deferred by some mechanism if the change is triggered by a rectangular selection action.

**Comment by zaksmolen** [comment]
Ugh, well the side effect of not calling `Changed()` around line 877 is that it doesn't update the caret for normal typing.

**Comment by goswinr** [comment]
@zaksmolen I could significantly improve the performance of typing in a rectangle selection  by using TextArea.PreviewTextInput event , then doing the block edit myself via a Stringbuilder and then call doc.Replace only once.

**Comment by zaksmolen** [comment]
I believe I have finally resolved this in pull request 269.
